tests/purs/failing/FoldableInstance9.out

```diff
@@ -0,0 +1,42 @@
+Error found:
+in module [33mFoldableInstance9[0m
+at tests/purs/failing/FoldableInstance9.purs:53:1 - 53:38 (line 53, column 1 - line 53, column 38)
+
+  The following data constructors and their subterms must use the type variable 'a' only as the last argument of a data type:
+
+    Constructor, Test1 has invalid 'a' usage at path(s):
+
+      _0_0
+      _0_1
+      _1_1
+      _2_0
+      _2_1
+      _3_1
+
+    Constructor, Test2 has invalid 'a' usage at path(s):
+
+      _0."all"_0
+      _0."all"_1
+
+    Constructor, Test5 has invalid 'a' usage at path(s):
+
+      _0."all"_0
+      _0."all"_1
```

Comment:
Two things: first, this error message misses quite a lot of invalid occurrences of `a`. Every single constructor has at least one, and there appear to be only a handful of *valid* occurrences in the entire declaration.

Second: the following approach is unprecedented in our error messages, but then again so is this path-based presentation so it seems we want an innovation either way. What if we aimed for something like this:

```
The following occurrences of 'a' prevent 'Foldable (Test f g h)' from being derived:

  16:  = Test1 (f a a a) (f Int a a) (f a a Int) (f Int a Int) (f Int Int Int)
                  ^ ^           ^       ^ ^             ^
  17:  | Test2 { all :: f a a a
                          ^ ^
  18:          , rights :: f Int a a
                                 ^
  19:          , lefts :: f a a Int
                            ^ ^
  20:          , middle :: f Int a Int
                                 ^
```

The biggest downside might be that if these messages are presented in a proportional-width font, they'd be next to useless. Maybe we could use a different format in the IDE interface for that reason.

This is something that could be done as a later refinement PR, so don't let it hold things up if it's a controversial idea.

Reply:
> Every single constructor has at least one, and there appear to be only a handful of valid occurrences in the entire declaration.

I think that's only the case because I haven't pushed an update for this test. Locally, there's more things listed.

Reply:
> Second: the following approach is unprecedented in our error messages, but then again so is this path-based presentation so it seems we want an innovation either way. What if we aimed for something like this:

Yeah... I've been wondering the same thing. The path-based representation does give you all the info you need in a succinct way, but it's not readable unless someone clicks on the link to the documentation repo (a page which doesn't exist yet).

I think the format you've shown above is what I'd prefer personally (though it means more work :smile:). While there may be some situations where the formatting isn't good, I think that will happen infrequently. So overall, a better tradeoff than the path-based one.

Reply:
Another way to workaround the proportional font issue is to color the problematic `a` usages a different color in the error message. Most terminals used by most users support colors.